compiler/rustc_mir/src/transform/lower_slice_len.rs

```diff
@@ -0,0 +1,144 @@
+        return;
+    };
+    let len_symbol = rustc_span::Symbol::intern("len");
+
+    let mut visitor = LowerLenCallVisitor { tcx, len_symbol, slice_lang_item_def_id };
+    visitor.visit_body(body);
+}
+
+struct LowerLenCallVisitor<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    len_symbol: rustc_span::Symbol,
+    slice_lang_item_def_id: DefId,
+}
+
+struct PatchInformation<'tcx> {
+    add_statement: Statement<'tcx>,
+    new_terminator_kind: TerminatorKind<'tcx>,
+}
+
+impl<'tcx> MutVisitor<'tcx> for LowerLenCallVisitor<'tcx> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+
+    fn visit_basic_block_data(&mut self, _block: BasicBlock, data: &mut BasicBlockData<'tcx>) {
+        // We do not need to call super here, so just work with it
+
+        let mut patch_found: Option<PatchInformation<'_>> = None;
+
+        let terminator = data.terminator();
+        match &terminator.kind {
+            TerminatorKind::Call {
+                func: Operand::Constant(box func_sig),
+                args,
+                destination: Some((dest, bb)),
+                cleanup: None,
+                from_hir_call: true,
+                ..
+            } => {
+                // some heuristics for fast rejection
+                if args.len() != 1 {
+                    return;
+                }
+                let (arg, _is_move) = match &args[0] {
+                    Operand::Copy(arg) => (*arg, false),
+                    Operand::Move(arg) => (*arg, true),
+                    _ => return,
+                };
+                let sig =
+                    if let Some(sig) = func_sig.literal.const_for_ty() { sig } else { return };
+                // match over function signature
+                match sig {
+                    ty::Const {
+                        ty,
+                        val:
+                            ty::ConstKind::Value(rustc_middle::mir::interpret::ConstValue::Scalar(
+                                rustc_middle::mir::interpret::Scalar::Int(s),
+                            )),
+                    } => {
+                        // more early rejections
+                        if s != &ty::ScalarInt::ZST || !ty.is_fn() {
+                            return;
+                        }
+                        match ty.kind() {
+                            ty::FnDef(def_id, _) => {
+                                let item_name = self.tcx.item_name(*def_id);
+                                let impl_name =
+                                    if let Some(impl_name) = self.tcx.impl_of_method(*def_id) {
+                                        impl_name
+                                    } else {
+                                        return;
+                                    };
+                                if item_name == self.len_symbol
+                                    && impl_name == self.slice_lang_item_def_id
+                                {
```

Comment:
I would personally prefer if `len` became lang item with a comment that describes that an optimization expects a certain behavior.

Reply:
Can you guide me on how to make a new language item? I found a definition in language_item_table!, but it looks like magic. Don't even understand how `slice_impl` works

Reply:
I think the line in that table should be something like `SliceLen, sym::slice_len, slice_len_fn, Target::Fn;` Then the lang item is called `slice_len_fn`. You need to use `#[cfg_attr(not(bootstrap), lang = "slice_len_fn")]` as attribute for the `len` function definition in libcore as the bootstrap compiler doesn't know about it.

Reply:
Oh, it's backwards and is taken from code and not some black magic. Ok, this should be enough for me